[3.10] Joomla.Showon.initialise API, and  make sure that showon initialised only once, fork of 18843

Fedik opened the pull request:
Pull Request for Issue #18843 and only for J3 .

### Summary of Changes

This is fork of #18843, with requested changes.
The code make sure that showon initialised only once, no mater how often Joomla.setUpShowon was called.
This is achieved with extra flag `showonInitialised` that sets while showon initialisation.


### Testing Instructions

Make sure `showon` still works.

Example, go to global config change "Debug Language" on/off, should downup/hides "Language Display" field.
Also the field "Language Display"  will have extra attribute `data-showon-initialised`


### Actual result BEFORE applying this Pull Request
showon works,
the showon field do not have `data-showon-initialised`



### Expected result AFTER applying this Pull Request
showon works,
the showon field do have `data-showon-initialised` attribute

cheesegrits committed:
Add an event to parse a form added to the DOM on the fly for showon.

cheesegrits committed:
Merge branch 'staging' into 3.8-showon

cheesegrits committed:
Changed using event to renaming setUpShowon as Joomla.setUpShowon.

Fedik committed:
Merge branch 'staging' into pull18843

Fedik committed:
Set up showon only once

richard67 commented:
@Fedik #18843 can be closed, right?

Fedik commented:
yeah, I have included changes from that pull

dgrammatiko commented:
@Fedik diffing the changes with the v4 version of show on I think you need to rename this to `Joomla.Showon.initilise` to be called like `Joomla.Showon.initilise(field)`. The initialise in J4 could be a function in the Class with this code (the refactoring should be straight forward):
https://github.com/joomla/joomla-cms/blob/d49e4614f299b5f158f6a585de6f9c0313013168/build/media_source/system/js/showon.es6.js#L32-L84

Fedik commented:
This only for J3, it should not go to J4,
there already event stuff for such thing:
https://github.com/joomla/joomla-cms/blob/d49e4614f299b5f158f6a585de6f9c0313013168/build/media_source/system/js/showon.es6.js#L215-L218

Only missed check for "double initialisation", I can do it later in another pull.

I do not think that this patch is much needed, but people asked :wink: 
It can be safely closed, depend what @HLeithner will decide

HLeithner commented:
@Fedik would be cool if we have the double initialization added in this pr too. After 2 tests we can merge it in to 3.10 (I think that makes more sense). I don't expect another 3.9 release (hopefully). It also introduces a new API endpoint so 3.10 is a better target.

Fedik commented:
> would be cool if we have the double initialization added in this pr too

it is here, that why I made this fork:
https://github.com/joomla/joomla-cms/blob/0f5956e99c8f859e0668b790a5e31629b964688b/media/jui/js/cms-uncompressed.js#L184-L189

I meant that for J4 will be need a separted PR.

> It also introduces a new API endpoint so 3.10 is a better target.

I can remove Joomla.setUpShowon and leave "event" $(document).on('subform-row-add'), I guess most people who need it already use in this way, so it will be "bug fix" :wink:

Fedik committed:
Backport joomla:updated, public Joomla.Showon.initilise

richard67 commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -209,31 +209,47 @@ Joomla = window.Joomla || {};
-					$elm.attr('data-showon', showon);
+			Joomla.Showon.initilise(document);
+
+            // Setup showon feature in the modified container
```
```suggestion
			// Setup showon feature in the modified container
```

Fedik commented:
Okay, so, I have backport Joomla 4 'joomla:updated' stuff for showon, now it can be more easy for transition.
And added `Joomla.Showon.initilise` as @dgrammatiko suggested, later I will make pull  for J4.

Testing the same, make sure `showon` still work.

Please someone change target branch to joomla-10, or I have to make new PR?

Fedik committed:
Update media/jui/js/cms-uncompressed.js

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Fedik committed:
correct target

Fedik committed:
Merge branch 'pull18843-2' of github.com:Fedik/joomla-cms into pull18843-2

HLeithner commented:
I change the branch for you.

But you can do this your self by adding the PR
![image](https://user-images.githubusercontent.com/1497730/104837590-dcb36000-58b5-11eb-81ee-dec066099c08.png)

Thanks for your PR, now we need some tests ;-)

Fedik commented:
well, I just broke PR :smile: 
need some time to fix

Fedik commented:
@HLeithner can I merge joomla:3.10-dev to my branch for sync? or I need make new

HLeithner committed:
Merge branch '3.10-dev' into pull18843-2

HLeithner commented:
I updated your branch you only need to pull from github again.

HLeithner commented:
but yes you can simple merge 3.10-dev into your branch

Fedik commented:
okay, thanks

richard67 commented:
@HLeithner Now the PR shows more changes than only those from @Fedik . It shows the changes which meanwhile have been made in the staging branch since the last upmerge of staging into 3-10-dev. So maybe it's easier just to redo the PR for 3.-10-dev?

Fedik commented:
@richard67 or we can wait when `staging` will be merged in to 3-10-dev :)

richard67 commented:
Yes, that will also work.

HLeithner commented:
I will upmerge staging

HLeithner committed:
Merge branch '3.10-dev' into pull18843-2

HLeithner commented:
so now it's ok again

richard67 commented:
Yes, looks ok now.

Fedik committed:
use same attribute name as in j4

Fedik committed:
Merge branch 'pull18843-2' of github.com:Fedik/joomla-cms into pull18843-2

Fedik commented:
I have updated testing instruction, and made same changes for joomla 4 #32069

dgrammatiko commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -195,35 +204,52 @@ Joomla = window.Joomla || {};
+		// Provide a public API
+		if (!Joomla.Showon) {
+			Joomla.Showon = {
+				initilise: setUpShowon
```
Ooops typo: `initialise `

Fedik replied:
fixed

Fedik committed:
typo

dgrammatiko commented:
I have tested this item :white_check_mark: successfully on 04e2d9de9fe18daca3a3cb70f4a66ab864505129<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

gostn commented:
I have tested this item :white_check_mark: successfully on 04e2d9de9fe18daca3a3cb70f4a66ab864505129<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

richard67 committed:
Merge branch '3.10-dev' into pull18843-2

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

zero-24 commented:
@Fedik do we need to add documentation for this change somewhere? And if so what should be documented?

Fedik commented:
@zero-24 good question, I have no idea ;) 
Do we have any documentation about `showon` API somwhere? if no, then maybe we need to add.

Technically and partly it is a new feature, to cover edge cases from #18843 , also see comment  https://github.com/joomla/joomla-cms/pull/18843#issuecomment-346930106

In J4 we already have events #16016 (I backported it partly here), that handle "dynamic" content changes

zero-24 commented:
Ok will merge than this heree for 3.10. thanks
